certora/specs/RevertsERC20.spec

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function allowance(address, address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+}
+
+// Check the revert conditions for the transfer function.
+rule transferRevertConditions(env e, address to, uint256 amount) {
+    uint256 balanceOfSenderBefore = balanceOf(e.msg.sender);
+    uint256 senderVotingPowerBefore = delegatedVotingPower(delegatee(e.msg.sender));
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(to));
+
+    // Safe require as it is verified in delegatedLTEqDelegateeVP.
+    require senderVotingPowerBefore >= balanceOfSenderBefore;
+    // Safe require because if delegatees are different the recipient's voting power excludes the holder's balance.
+    require delegatee(to) != delegatee(e.msg.sender) => recipientVotingPowerBefore <= totalSupply() - balanceOfSenderBefore;
```

Comment:
Let's rewrite it in terms of an invariant (or multiple invariants) that we think hold

Reply:
Here I think the previous require helps, so we can decompose it into:
- `delegatee(to) != delegatee(e.msg.sender) => recipientVotingPowerBefore + senderVotingPowerBefore <= sumOfVotingPower`
- `sumOfVotingPower <= totalSupply()`

Reply:
I think it suffices to say this follow from the lemmas:
- `totalSupplyIsVirtualSumOfVotes()` (see https://github.com/morpho-org/morpho-token/pull/104/commits/7db9d33e3c25fbec5380c17ff872e57edd4fa93b);
- `delegatedLTEqDelegateeVP()`;
- `totalSupplyIsSumOfVirtualVotingPower()`. 

Note that `sumOfVotingPower` isn't accessible (defined in a file that needs munging).

Reply:
Done in https://github.com/morpho-org/morpho-token/pull/98/commits/75370ef69778f70d168ff454e7dd6bb1291bff49.

Reply:
Reopening because of the "optimization" stated [here](https://github.com/morpho-org/morpho-token/pull/98#discussion_r1878066033), in case you have missed it. But we can move along without it if you prefer

Reply:
It was actually implemented already, my bad